pymongo/bulk.py

```diff
@@ -267,6 +267,14 @@ def _execute_command(self, generator, write_concern, session,
         # sock_info.write_command.
         sock_info.validate_session(client, session)
         while run:
+            # Check to see if this is the last run by peeking.
+            # On the last run we use the final write concern.
+            try:
+                next_run = next(generator)
+                generator = chain([next_run], generator)
+            except StopIteration:
+                write_concern = final_write_concern or write_concern
```

Comment:
A "run" is a sequence of the same operation type, eg `bulk_write([UpdateOne({}), InsertOne({}), InsertOne({})])` will have one run with one Update and then a second run with two inserts. I don't think this implementation handles the case where the last run is batched into multiple commands. In that case, all the commands must use w:1 writeConcern expect the last command.

I think the actual condition would be:
```
if last_run and last_command_in_run:
    use final_write_concern
else:
    use write_concern
```

However, the point at which we batch split happens _after_ encoding the writeConcern field to the wire protocol message. So this is somewhat of a chicken and egg problem.

Reply:
I think the latest commit addresses this